altair/utils/schemapi.py

```diff
@@ -69,83 +126,181 @@ def validate_jsonschema(spec, schema, rootschema=None):
-            most_relevant_errors = parent.context or []
-            if lowest_level.validator == "enum":
-                # There might be other possible enums which are allowed, e.g. for
-                # the "timeUnit" property of the "Angle" encoding channel. These do not
-                # necessarily need to be in the same branch of this tree of errors that
-                # we traversed down to the lowest level. We therefore gather
-                # all enums in the leaves of the error tree.
-                enum_errors = _get_all_lowest_errors_with_validator(
-                    top_level_error, validator="enum"
-                )
-                # Remove errors which already exist in enum_errors
-                enum_errors = [
-                    err
-                    for err in enum_errors
-                    if err.message not in [e.message for e in most_relevant_errors]
-                ]
-                most_relevant_errors = most_relevant_errors + enum_errors
-
-    # This should never happen but might still be good to test for it as else
-    # the original error would just slip through without being raised
-    if len(most_relevant_errors) == 0:
-        raise Exception("Could not determine the most relevant errors") from errors[0]
-
-    return most_relevant_errors
-
-
-def _get_all_lowest_errors_with_validator(
-    error: jsonschema.ValidationError, validator: str
-) -> List[jsonschema.ValidationError]:
-    matches: List[jsonschema.ValidationError] = []
-    if error.context:
-        for err in error.context:
-            if err.context:
-                matches.extend(_get_all_lowest_errors_with_validator(err, validator))
-            elif err.validator == validator:
-                matches.append(err)
-    return matches
+            leaves.append(err)
+    return leaves
+
+
+def _subset_to_most_specific_json_paths(
+    errors_by_json_path: GroupedValidationErrors,
+) -> GroupedValidationErrors:
+    """Removes key (json path), value (errors) pairs where the json path is fully
+    contained in another json path. For example if `errors_by_json_path` has two
+    keys, `$.encoding.X` and `$.encoding.X.tooltip`, then the first one will be removed
+    and only the second one is returned. This is done under the assumption that
+    more specific json paths give more helpful error messages to the user.
+    """
+    errors_by_json_path_specific: GroupedValidationErrors = {}
+    for json_path, errors in errors_by_json_path.items():
+        if not _contained_at_start_of_one_of_other_values(
+            json_path, list(errors_by_json_path.keys())
+        ):
+            errors_by_json_path_specific[json_path] = errors
+    return errors_by_json_path_specific
+
+
+def _contained_at_start_of_one_of_other_values(x: str, values: Sequence[str]) -> bool:
+    # Does not count as "contained at start of other value" if the values are
+    # the same. These cases should be handled separately
```

Comment:
So if the values are exactly the same are they deduplicated elsewhere instead? Maybe in the function just below?

Reply:
Exactly, they are deduplicated by message on line 210